File: ios/StreamaxCore.swift
import Foundation

/// Swift-flavored wrapper around the Rust streamax-core C ABI.
/// Single-threaded — call from one queue at a time.
final class StreamaxCore {
    enum Event {
        case videoConfig(annexB: Data, width: Int, height: Int)
        case videoFrame(annexB: Data, ptsMs: UInt32, isKeyframe: Bool)
        case audioConfig(config: Data, sampleRate: Int, channels: Int, objectType: Int)
        case audioFrame(data: Data, ptsMs: UInt32)
        case error(String)
    }

    private var handle: OpaquePointer?

    init() {
        handle = OpaquePointer(streamax_demuxer_new())
    }

    deinit {
        if let h = handle { streamax_demuxer_free(UnsafeMutableRawPointer(h)) }
    }

    func reset() {
        guard let h = handle else { return }
        streamax_demuxer_reset(UnsafeMutableRawPointer(h))
    }

    func append(_ chunk: Data) {
        guard let h = handle, !chunk.isEmpty else { return }
        chunk.withUnsafeBytes { raw in
            if let base = raw.bindMemory(to: UInt8.self).baseAddress {
                streamax_demuxer_append(UnsafeMutableRawPointer(h), base, chunk.count)
            }
        }
    }

    func nextEvent() -> Event? {
        guard let h = handle else { return nil }
        var ev = StreamaxEvent()
        let got = streamax_demuxer_next_event(UnsafeMutableRawPointer(h), &ev)
        if got == 0 { return nil }

        // Pointer is valid until the next FFI call — copy out now.
        let data: Data = ev.data_len > 0 && ev.data != nil
            ? Data(bytes: ev.data!, count: ev.data_len)
            : Data()

        switch ev.kind {
        case STREAMAX_EVENT_VIDEO_CONFIG:
            return .videoConfig(annexB: data, width: Int(ev.width), height: Int(ev.height))
        case STREAMAX_EVENT_VIDEO_FRAME:
            return .videoFrame(annexB: data, ptsMs: ev.pts_ms, isKeyframe: ev.is_keyframe == 1)
        case STREAMAX_EVENT_AUDIO_CONFIG:
            return .audioConfig(config: data, sampleRate: Int(ev.sample_rate),
                                channels: Int(ev.channels), objectType: Int(ev.object_type))
        case STREAMAX_EVENT_AUDIO_FRAME:
            return .audioFrame(data: data, ptsMs: ev.pts_ms)
        case STREAMAX_EVENT_ERROR:
            return .error(String(data: data, encoding: .utf8) ?? "demuxer error")
        default:
            return nil
        }
    }
}

File: android/app/src/main/java/com/example/streamax/StreamaxCore.kt
package com.example.streamax

/**
 * Kotlin wrapper for the Rust streamax-core native library.
 *
 * Build pipeline:
 *   1. Build `streamax-core` for Android (see streamax-core/build.sh).
 *   2. Copy libstreamax_core.so + the JNI bridge into `src/main/jniLibs/<abi>/`.
 *
 * The JNI bridge is in `app/src/main/cpp/streamax_jni.c`; it links against
 * libstreamax_core.so and exposes the methods declared here.
 *
 * Single-threaded — create one instance per stream.
 */
class StreamaxCore : AutoCloseable {

    sealed class Event {
        data class VideoConfig(val annexB: ByteArray, val width: Int, val height: Int) : Event()
        data class VideoFrame(val annexB: ByteArray, val ptsMs: Int, val isKeyframe: Boolean) : Event()
        data class AudioConfig(
            val config: ByteArray,
            val sampleRate: Int,
            val channels: Int,
            val objectType: Int,
        ) : Event()
        data class AudioFrame(val data: ByteArray, val ptsMs: Int) : Event()
        data class Error(val message: String) : Event()
    }

    private var handle: Long = nativeNew()

    fun reset() {
        if (handle != 0L) nativeReset(handle)
    }

    fun append(chunk: ByteArray, length: Int) {
        if (handle != 0L && length > 0) nativeAppend(handle, chunk, length)
    }

    fun nextEvent(): Event? {
        if (handle == 0L) return null
        val ev = nativeNextEvent(handle) ?: return null
        return when (ev.kind) {
            KIND_VIDEO_CONFIG -> Event.VideoConfig(ev.data, ev.width, ev.height)
            KIND_VIDEO_FRAME -> Event.VideoFrame(ev.data, ev.ptsMs, ev.isKeyframe)
            KIND_AUDIO_CONFIG -> Event.AudioConfig(ev.data, ev.sampleRate, ev.channels, ev.objectType)
            KIND_AUDIO_FRAME -> Event.AudioFrame(ev.data, ev.ptsMs)
            KIND_ERROR -> Event.Error(String(ev.data, Charsets.UTF_8))
            else -> null
        }
    }

    override fun close() {
        if (handle != 0L) {
            nativeFree(handle)
            handle = 0L
        }
    }

    private external fun nativeNew(): Long
    private external fun nativeFree(handle: Long)
    private external fun nativeReset(handle: Long)
    private external fun nativeAppend(handle: Long, data: ByteArray, length: Int)
    private external fun nativeNextEvent(handle: Long): NativeEvent?

    private class NativeEvent {
        @JvmField var kind: Int = 0
        @JvmField var data: ByteArray = ByteArray(0)
        @JvmField var ptsMs: Int = 0
        @JvmField var isKeyframe: Boolean = false
        @JvmField var width: Int = 0
        @JvmField var height: Int = 0
        @JvmField var sampleRate: Int = 0
        @JvmField var channels: Int = 0
        @JvmField var objectType: Int = 0
    }

    companion object {
        private const val KIND_VIDEO_CONFIG = 1
        private const val KIND_VIDEO_FRAME = 2
        private const val KIND_AUDIO_CONFIG = 3
        private const val KIND_AUDIO_FRAME = 4
        private const val KIND_ERROR = 99

        init { System.loadLibrary("streamax_jni") }
    }
}
